File: principles-strategy-observer/upgrade-without-change.c
#include <stddef.h>
#include <stdio.h>
#include <string.h>

const void *mymax(
    const void *base, size_t nmemb, size_t size,
    int (*compar)(const void *, const void *)) {
        if (nmemb == 0) return NULL;

        const char *max = (const char*) base;
        const char *current = max;

        for (int i = 1; i < nmemb; i++) {
            current += size;
            if (compar(current, max) > 0) {
                max = current;
            }
        }

        return max;
    }

int gt_int(const void *first, const void *second) {
    int a = *(const int*) first;
    int b = *(const int*) second;
    if (a > b) return 1;
    if (a < b) return -1;
    return 0;
}

int gt_char(const void *first, const void *second) {
    const char a = *(const char*) first;
    const char b = *(const char*) second;
    if (a > b) return 1;
    if (a < b) return -1;
    return 0;
}

int gt_str(const void *first, const void *second) {
    const char *a = *(const char* const*) first;
    const char *b = *(const char* const*) second;
    return strcmp(a, b);
}

int main() {
    int arr_int[] = { 1, 3, 5, 7, 4, 6, 9, 2, 0 };
    char arr_char[]="Suncana strana ulice";
    const char* arr_str[] = { "Gle", "malu", "vocku", "poslije", "kise", "Puna", "je", "kapi", "pa", "ih", "njise"};

    const int* int_max = (const int*) mymax(&arr_int, 9, sizeof(arr_int[0]), gt_int);
    printf("Int max: %d\n", *int_max);

    const char* char_max = (const char*) mymax(&arr_char, strlen(arr_char), sizeof(arr_char[0]), gt_char);
    printf("Char max: %c\n", *char_max);

    const char* str_max = *(const char* const*) mymax(arr_str, 11, sizeof(arr_str[0]), gt_str);
    printf("Str max: %s\n", str_max);
}

File: principles-strategy-observer/upgrade-without-change.py
def mymax(iterable, key=lambda x: x):
    max_x=max_key=None

    for x in iterable:
        if not max_x or key(x) > max_key:
            max_x = x
            max_key = key(x)

    return max_x

def string_list_test():
    string_list = ["aaa", "aac", "z", "aza", "zz"]
    string_key = lambda x: len(x)
    
    print(mymax(string_list, key=string_key))
    
def test_default_argument():
    max_int = mymax([1, 3, 5, 7, 4, 6, 9, 2, 0])
    max_char = mymax("Suncana strana ulice")
    max_string = mymax([
    "Gle", "malu", "vocku", "poslije", "kise",
    "Puna", "je", "kapi", "pa", "ih", "njise"])
    
    print(f"Max int: {max_int}")
    print(f"Max char: {max_char}")
    print(f"Max string: {max_string}")
    
def test_dictionary():
    D={'burek':8, 'buhtla':5}
    max_val = mymax(D, D.get)
    print(f"Max val in dict: {max_val}") 
    
def test_tuples():
    tuples = [("name", "surname"), ("john", "doe"), ("ime", "prezime"), ("z", "z")]
    max_name = mymax(tuples)
    print(f"Max name is: {max_name}")

if __name__ == "__main__":
    test_tuples()